test/PushNotificationTests/Shared.cpp

```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+#include <pch.h>
+#include <TestDef.h>
+#include "Shared.h"
+
+using namespace WEX::Common;
+using namespace WEX::Logging;
+using namespace WEX::TestExecution;
+
+using namespace winrt;
+using namespace winrt::Windows::ApplicationModel;
+using namespace winrt::Windows::ApplicationModel::Activation;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::Management::Deployment;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::System;
+
+namespace Test::PushNotifications
+{
+    StorageFile CreateDocFile(std::wstring filename)
+    {
+        return KnownFolders::DocumentsLibrary().CreateFileAsync(filename,
+            CreationCollisionOption::OpenIfExists).get();
+    }
+
+    StorageFile OpenDocFile(std::wstring filename)
+    {
+        return KnownFolders::DocumentsLibrary().GetFileAsync(filename).get();
+    }
+
+    wil::unique_handle Execute(const std::wstring& command, const std::wstring& args,
+        const std::wstring& directory)
+    {
+        SHELLEXECUTEINFO ei{};
+        ei.cbSize = sizeof(SHELLEXECUTEINFO);
+        ei.fMask = SEE_MASK_NOCLOSEPROCESS | SEE_MASK_DOENVSUBST;
+        ei.lpFile = command.c_str();
+        ei.lpParameters = args.c_str();
+        ei.lpDirectory = directory.c_str();
+
+        if (!ShellExecuteEx(&ei))
+        {
+            auto lastError = GetLastError();
+            VERIFY_WIN32_SUCCEEDED(lastError);
+        }
+
+        wil::unique_handle process{ ei.hProcess };
+        return process;
+    }
+
+    void RunCertUtil(const std::wstring& path, bool removeCert)
+    {
+        std::wstring action = (removeCert ? L"-delstore" : L"-addstore");
+        std::wstring args{ action + L" TrustedPeople " + path };
+        // VERIFY_ARE_NOT_EQUAL(args, args);
+        auto process = Execute(L"%SystemRoot%\\system32\\certutil.exe",
+            args, g_deploymentDir);
+
+        // Wait for the cer to be installed.
+        auto waitResult = WaitForSingleObject(process.get(), c_phaseTimeout);
+        if (waitResult != WAIT_OBJECT_0)
+        {
+            if (waitResult == WAIT_FAILED) VERIFY_WIN32_SUCCEEDED(GetLastError());
+            VERIFY_ARE_EQUAL(waitResult, WAIT_OBJECT_0);
+        }
+
+        // Make sure the exitcode for the tool is success.
+        DWORD exitCode{};
+        THROW_IF_WIN32_BOOL_FALSE(GetExitCodeProcess(process.get(), &exitCode));
+        VERIFY_ARE_EQUAL(exitCode, 0);
+    }
+
+    void InstallPackage(const std::wstring& packagePath)
```

Comment:
Delete this and use `TP::AddPackage()` in test/inc/ProjectReunion.Test.Package.h

Even better, add your test package(s) there and just use that
Clone/fork `DynamicDependencyDataStore` (aka PRmain for testing) in `test/inc/ProjectReunion.Test.Package.h` - find references to `DynamicDependencyDataStore` and make your own equivalent(s).

Doing that obsoletes the need for `InstallPackage()` and `UninstallPackage()`

Reply:
The issue with the AddPackage() function is that it assumes the BuildOutput dir is the same name as the package that needs to be installed as well it being a .msix package. I cannot install VCLibs.appx because of this. Can we make AddPackage generic to other package types?

Reply:
>Can we make AddPackage generic to other package types?
No, and I'm not sure it's needed. See below

> AddPackage()...assumes the BuildOutput dir is the same name as the package that needs to be installed
Yes, that assumes you're installing a package built by ProjectReunion.sln so it does things to find the package making assumptions about build structure

```
    inline void AddPackage(PCWSTR packageDirName, PCWSTR packageFullName)
    {
        // Build the target package's .msix filename. It's under the Solution's $(OutDir)
        // NOTE: It could live in ...\Something.msix\... or ...\Something\...
        auto solutionOutDirPath = ::Test::FileSystem::GetSolutionOutDirPath();
        //
        // Look in ...\Something.msix\...
        auto msix(solutionOutDirPath);
        msix /= packageDirName;
        msix += L".msix";
        msix /= packageDirName;
        msix += L".msix";
...
```

If you want a generic install-any-package (like VCLibs.appx) you can directly use the PackageManager API. Piece of cake via C++/WinRT

```
        std::wstring msix{ L"c:\foo\bar.msix" };
        auto msixUri = winrt::Windows::Foundation::Uri(msix.c_str());

        winrt::Windows::Management::Deployment::PackageManager packageManager;
        auto options{ winrt::Windows::Management::Deployment::DeploymentOptions::None };
        auto deploymentResult{ packageManager.AddPackageAsync(msixUri, nullptr, options).get() };
        VERIFY_SUCCEEDED(deploymentResult.ExtendedErrorCode(), WEX::Common::String().Format(L"AddPackageAsync('%s') = 0x%0X %s", packageFullName, deploymentResult.ExtendedErrorCode(), deploymentResult.ErrorText().c_str()));
```

That `.get()` on `pkgmgr.BlahAsync().get()` makes it a synchronous call so easy peasy.

That's not your problem.

>I cannot install VCLibs.appx 

Do you need to? (Does VS install it when VS is installed? Is it present in Helix environments?)

If you do, **where do you get it?**

`std::wstring msix{ L"c:\foo\vclibs.appx" };` is the question. Where are/would you get VCLibs from?

And do you need it?

Have you seen [Use the Universal C Runtime #888](https://github.com/microsoft/ProjectReunion/pull/888)? That lets Microsoft.ProjectReunion.dll etc work w/o needing the VCLibs package or VC CRT redist (about the only option to make DLL meet both scenarios).

Reply:
My issue is that when I build my MSIX packaged test app, the name of my package is not PushNotificationsTestAppPackage.msix, it is PushNotificationsTestAppPackage_1.0.0.0_x64. Am I supposed to do change the build file to be named PushNotificationsTestAppPackage on build?

Reply:
Ack! That's provided by `PushNotificationsTestAppPackage.wapproj`

So it generates a file in the project's outdir named <project>_<version>_<arch>.appx?

Probably .props magic for .wapproj support as your manifest doesn't specify architecture

```
<Identity
    Name="PushNotificationsTestAppPackage"
    Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US"
    Version="1.0.0.0" />
```

OK how about this

```
    constexpr PCWSTR GetCurrentArchitectureAsString()
    {
    #if defined(_M_X64)
        return L"x64";
    #elif defined(_M_IX86)
        return L"x86";
    #elif defined(_M_ARM64)
        return L"arm64";
    #elif defined(_M_ARM)
        return L"arm";
    #else
    #   error "Unknown processor architecture"
    #endif
    }

    namespace WapProj
    {
        inline void AddPackage(PCWSTR packageDirName, PCWSTR packageName, PCWSTR packageVersion)
        {
            auto msix(::Test::FileSystem::GetSolutionOutDirPath());
            msix /= packageDirName;
            msix /= packageName;
            msix += "_";
            msix += packageVersion;
            msix += "_";
            msix += GetCurrentArchitectureAsString();
            msix += ".appx";
            auto msixUri = winrt::Windows::Foundation::Uri(msix.c_str());

            // Install the package
            winrt::Windows::Management::Deployment::PackageManager packageManager;
            auto options{ winrt::Windows::Management::Deployment::DeploymentOptions::None };
            auto deploymentResult{ packageManager.AddPackageAsync(msixUri, nullptr, options).get() };
            VERIFY_SUCCEEDED(deploymentResult.ExtendedErrorCode(), WEX::Common::String().Format(L"AddPackageAsync('%s') = 0x%0X %s", msix.c_str(), deploymentResult.ExtendedErrorCode(), deploymentResult.ErrorText().c_str()));
        }

        inline void AddPackage(PCWSTR packageDirName, PCWSTR packageVersion)
        {
            AddPackage(packageDirName, packageDirName, packageVersion);
        }
    }
```

and call it as `TP::WapProj::InstallPackage(L"PushNotificationsTestAppPackage", L"1.0.0.0")`

?

Reply:
It is an  .msix but it does need the pkg. version + architecture.